packages/common/infra/src/op/consumer.ts

```diff
@@ -0,0 +1,188 @@
+import EventEmitter2 from 'eventemitter2';
+import {
+  defer,
+  from,
+  fromEvent,
+  Observable,
+  of,
+  share,
+  take,
+  takeUntil,
+} from 'rxjs';
+
+import {
+  AutoMessageHandler,
+  type CallMessage,
+  fetchTransferables,
+  type MessageHandlers,
+  type ReturnMessage,
+  type SubscribeMessage,
+  type SubscriptionCompleteMessage,
+  type SubscriptionErrorMessage,
+  type SubscriptionNextMessage,
+} from './message';
+import type { OpInput, OpNames, OpOutput, OpSchema } from './types';
+
+interface OpCallContext {
+  signal: AbortSignal;
+}
+
+export type OpHandler<Ops extends OpSchema, Op extends OpNames<Ops>> = (
+  payload: OpInput<Ops, Op>[0],
+  ctx: OpCallContext
+) =>
+  | OpOutput<Ops, Op>
+  | Promise<OpOutput<Ops, Op>>
+  | Observable<OpOutput<Ops, Op>>;
+
+export class OpConsumer<Ops extends OpSchema> extends AutoMessageHandler {
+  private readonly eventBus = new EventEmitter2();
+
+  private readonly registeredOpHandlers = new Map<
+    OpNames<Ops>,
+    OpHandler<Ops, any>
+  >();
+
+  private readonly processing = new Map<string, AbortController>();
+
+  override get handlers() {
+    return {
+      call: this.handleCallMessage,
+      cancel: this.handleCancelMessage,
+      subscribe: this.handleSubscribeMessage,
+      unsubscribe: this.handleCancelMessage,
+    };
+  }
+
+  private readonly handleCallMessage: MessageHandlers['call'] = async msg => {
+    const abortController = new AbortController();
+    this.processing.set(msg.id, abortController);
+
+    this.eventBus.emit(`before:${msg.name}`, msg.payload);
+    this.ob$(msg, abortController.signal)
+      .pipe(take(1))
+      .subscribe({
+        next: data => {
+          this.eventBus.emit(`after:${msg.name}`, msg.payload, data);
+          const transferables = fetchTransferables(data);
+          this.port.postMessage(
+            {
+              type: 'return',
+              id: msg.id,
+              data,
+            } satisfies ReturnMessage,
+            { transfer: transferables }
+          );
+        },
+        error: error => {
+          this.port.postMessage({
+            type: 'return',
+            id: msg.id,
+            error: error as Error,
+          } satisfies ReturnMessage);
+        },
+        complete: () => {
+          this.processing.delete(msg.id);
```

Comment:
should we send finish message ? Otherwise the caller's promise will never be fulfilled ？

Reply:
the path should always fall into next or error, otherwise the business logic get problem